### DISCUSSION TITLE
How to get texture square from an atlas texture?

### DISCUSSION THREAD

**Discussion by lokiiarora** [donmccurdy/glTF-Transform#807] [Q&A] [answered]
First of all, huge thanks taking the herculean effort of maintaining this library 🙏🏻
Has been a huge boon for us at Flipkart :) !
We are looking to make a post process function which can work on texture atlases that we generate internally and we've been struggling to figure out how to get the pixels used in a particular `textureInfo` of a `primitive`.

Let me know if more details are required on the same

**Comment by donmccurdy** [comment, marked as answer]
Hi @lokiiarora — really glad to hear it's been helpful, thanks!

The contents of `texture.getImage()` is the binary contents of the image, in whatever format. I'm currently using [`ndarray-pixels`](https://github.com/donmccurdy/ndarray-pixels) for pixel operations, because it's light, works across Web and Node.js, and allows arbitrary operations. 

General usage, for example:

```javascript
import { getPixels, savePixels } from 'ndarray-pixels';

const prim; // Primitive
const material = prim.getMaterial();
const texture = material.getBaseColorTexture();

// read
let image = texture.getImage();
const pixels = await getPixels(image, texture.getMimeType()); // Uint8Array -> ndarray

// modify
for (let i = 0; i < pixels.shape[0]; ++i) {
  for (let j = 0; j < pixels.shape[1]; ++j) {
    pixels.set(i, j, 255);
  }
}

// write
image = await savePixels(pixels, texture.getMimeType()); // ndarray -> Uint8Array
texture.setImage(image);
```

Here are specific examples of how I've used it, internally:

https://github.com/donmccurdy/glTF-Transform/blob/8062f29382b93877485273cd5d02c4279ef97468/packages/functions/src/metal-rough.ts#L95-L107

https://github.com/donmccurdy/glTF-Transform/blob/8062f29382b93877485273cd5d02c4279ef97468/packages/functions/src/utils.ts#L34-L59


***

However, `ndarray-pixels`:

1. doesn't support some image formats, like KTX2
2. isn't as fast or memory-efficient as Node.js native libraries (like [sharp](https://sharp.pixelplumbing.com/)) or WASM

... so depending on what you're doing, you may find that there are better options, and you should be able to use those with glTF Transform too. Most image-processing libraries should be able to accept the Uint8Array returned by `texture.getImage()` I think.

**Reply by donmccurdy** [reply]
> how to get the pixels used in a particular textureInfo of a primitive.

Or are you asking about the exact pixels that a particular primitive uses within the texture? You can do that but it does require looping over every triangle of the primitive, and every pixel within the UV space of those triangles, and is more complex. In many cases I think I'd want to use a graphics API for that, but it might help to know more what you're trying to do.

**Reply by lokiiarora** [reply]
I'm asking more along the lines of exact pixels used by a primitive in6case of usage of an extension like `KHR_Texture_Transform`

Is there a particular way i need to transform the UV coordinates to match the pixel?
The idea is to land a tile of the image that matches the bounds of UV of the primitive.
While I've done that, check whether the texture has a singular pixel representation and just eliminate that and remap UV for the entire texture (optional for now).

**Reply by donmccurdy** [reply]
If you've got the texture transform data from the texture already:

```javascript
const textureInfo = material.getBaseColorTextureInfo();
const transform = textureInfo.getExtension('KHR_texture_transform');
transform.getOffset(); // → [ 100, 100 ]
transform.getRotation(); // → 1.717
transform.getScale(); // → [ 2, 2 ]
```

...then you would need to use those transforms as in the [KHR_texture_transform specified math](https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_texture_transform/README.md) to transform the UVs — I use gl-matrix for most math operations internally, it should be fine for the matrix operations linked.

The part about visiting every pixel has one main wrinkle... graphics APIs have minFilter/magFilter settings that affect texture sampling, and nearby pixels can bleed into the UV island somewhat from that. Not sure I'd want to try to 'emulate' that with custom code. But probably nearest-pixel filtering would be fine if you just need to detect solid-colored primitives.

edit: I noticed the title says "texture square" - do you know for sure that each primitive's UVs are within a bounding box, and no other primitives are within that box? that probably makes it quite a bit easier if so...

**Reply by lokiiarora** [reply]
As of now, I am not sure.
I wanted to implement this as an idea, I have a sample model, I'll check within it for now !
Thanks for the math, I was looking for that, will share updates on this thread once experimented
